File: queries/tests/node/test_tags.js
const { describe, test } = require("node:test");
const assert = require("node:assert/strict");
const {
  getParser,
  getLanguage,
  loadQuery,
  runQueryMatches,
} = require("./helpers");

describe("tags.scm", () => {
  const parser = getParser();
  const lang = getLanguage();
  const query = loadQuery(lang, "tags.scm");
  const matches = runQueryMatches(lang, query, parser, "test_tags.cj");

  function matchesFor(role) {
    return matches.filter((match) =>
      match.captures.some((capture) => capture.name === role)
    );
  }

  function namesFor(role) {
    return new Set(matchesFor(role).map((match) => {
      const names = match.captures.filter((capture) => capture.name === "name");
      assert.equal(names.length, 1, `${role} match must contain exactly one @name`);
      return names[0].text;
    }));
  }

  function assertRoleNodeTypes(role, expectedTypes) {
    const allowed = new Set(expectedTypes);
    for (const match of matchesFor(role)) {
      const roles = match.captures.filter((capture) => capture.name === role);
      assert.equal(roles.length, 1, `${role} match must contain one role capture`);
      assert.ok(
        allowed.has(roles[0].type),
        `${role} must capture an outer definition node, got ${roles[0].type}`
      );
    }
  }

  test("every tag match pairs an outer definition with one @name", () => {
    for (const match of matches) {
      const roles = match.captures.filter((capture) =>
        capture.name.startsWith("definition.") || capture.name.startsWith("reference.")
      );
      const names = match.captures.filter((capture) => capture.name === "name");
      assert.equal(roles.length, 1);
      assert.equal(names.length, 1);
      const sameNode = roles[0].type === names[0].type &&
        JSON.stringify([roles[0].startPosition, roles[0].endPosition]) ===
          JSON.stringify([names[0].startPosition, names[0].endPosition]);
      assert.equal(
        sameNode,
        false,
        "the role capture must not be the name node (the legacy false-green shape)"
      );
    }
  });

  test("type definitions use protocol-compliant names and outer nodes", () => {
    assert.deepEqual(namesFor("definition.class"), new Set(["MyClass", "WithProperty", "Calculator"]));
    assert.deepEqual(namesFor("definition.struct"), new Set(["MyStruct"]));
    assert.deepEqual(namesFor("definition.interface"), new Set(["MyInterface"]));
    assert.deepEqual(namesFor("definition.enum"), new Set(["MyEnum"]));
    assertRoleNodeTypes("definition.class", ["classDefinition"]);
    assertRoleNodeTypes("definition.struct", ["structDefinition"]);
    assertRoleNodeTypes("definition.interface", ["interfaceDefinition"]);
    assertRoleNodeTypes("definition.enum", ["enumDefinition"]);
  });

  test("functions, main, init and operators tag their outer definitions", () => {
    const names = namesFor("definition.function");
    for (const expected of ["myFunction", "genericFunc", "classMethod", "main", "init", "=="]) {
      assert.ok(names.has(expected), `missing function tag ${expected}`);
    }
    assertRoleNodeTypes("definition.function", [
      "functionDefinition",
      "operatorFunctionDefinition",
      "mainDefinition",
      "init",
    ]);
  });

  test("enum constructors, aliases and properties use @name", () => {
    assert.deepEqual(namesFor("definition.constant"), new Set(["Value1", "Value2"]));
    assert.ok(namesFor("definition.type").has("MyTypeAlias"));
    assert.ok(namesFor("definition.property").has("value"));
    assertRoleNodeTypes("definition.constant", ["enumConstructor"]);
    assertRoleNodeTypes("definition.type", ["typeAlias"]);
    assertRoleNodeTypes("definition.property", ["propertyDefinition"]);
  });
});

File: tests/python/queries/test_tags.py
#!/usr/bin/env python3
"""Verify tags.scm against the shared Cangjie query fixture."""

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from helpers import get_parser_and_language, load_query, run_query_matches


class TestTags(unittest.TestCase):
    """Verify Tree-sitter's role + @name tag protocol, not raw captures only."""

    @classmethod
    def setUpClass(cls):
        cls.parser, cls.lang = get_parser_and_language()
        cls.query = load_query(cls.lang, "tags.scm")
        cls.matches = run_query_matches(
            cls.lang, cls.query, cls.parser, "test_tags.cj"
        )

    def _matches_for(self, role):
        return [captures for _, captures in self.matches if role in captures]

    def _names_for(self, role):
        names = set()
        for captures in self._matches_for(role):
            self.assertIn("name", captures)
            self.assertEqual(len(captures["name"]), 1)
            names.add(captures["name"][0][0])
        return names

    def _assert_role_types(self, role, expected):
        for captures in self._matches_for(role):
            self.assertEqual(len(captures[role]), 1)
            self.assertIn(captures[role][0][1], expected)

    def test_every_match_pairs_outer_definition_and_name(self):
        for _, captures in self.matches:
            roles = [name for name in captures if name.startswith(("definition.", "reference."))]
            self.assertEqual(len(roles), 1)
            self.assertIn("name", captures)
            self.assertEqual(len(captures["name"]), 1)
            role_capture = captures[roles[0]][0]
            name_capture = captures["name"][0]
            same_node = (
                role_capture[1] == name_capture[1]
                and (role_capture[2], role_capture[3])
                == (name_capture[2], name_capture[3])
            )
            self.assertFalse(same_node)

    def test_type_definitions(self):
        self.assertEqual(self._names_for("definition.class"), {"MyClass", "WithProperty", "Calculator"})
        self.assertEqual(self._names_for("definition.struct"), {"MyStruct"})
        self.assertEqual(self._names_for("definition.interface"), {"MyInterface"})
        self.assertEqual(self._names_for("definition.enum"), {"MyEnum"})
        self._assert_role_types("definition.class", {"classDefinition"})
        self._assert_role_types("definition.struct", {"structDefinition"})
        self._assert_role_types("definition.interface", {"interfaceDefinition"})
        self._assert_role_types("definition.enum", {"enumDefinition"})

    def test_function_like_definitions(self):
        names = self._names_for("definition.function")
        for expected in {"myFunction", "genericFunc", "classMethod", "main", "init", "=="}:
            self.assertIn(expected, names)
        self._assert_role_types(
            "definition.function",
            {"functionDefinition", "operatorFunctionDefinition", "mainDefinition", "init"},
        )

    def test_other_definitions(self):
        self.assertEqual(self._names_for("definition.constant"), {"Value1", "Value2"})
        self.assertIn("MyTypeAlias", self._names_for("definition.type"))
        self.assertIn("value", self._names_for("definition.property"))
        self._assert_role_types("definition.constant", {"enumConstructor"})
        self._assert_role_types("definition.type", {"typeAlias"})
        self._assert_role_types("definition.property", {"propertyDefinition"})


if __name__ == "__main__":
    unittest.main(verbosity=2)
